### ISSUE TITLE
Cloud-Init which is gzipped won't work

### ISSUE THREAD

**Issue by stffabi** [rancher/os#2391]
**RancherOS Version: (ros os version)**
1.4.0

**Where are you running RancherOS? (docker-machine, AWS, GCE, baremetal, etc.)**
Exoscale, CloudStack, all cloud-init providers which fetch userdata

Accordingly to [cloud-init documentation](http://cloudinit.readthedocs.io/en/latest/topics/format.html#gzip-compressed-content) the content can be gzip compressed. There are some providers which will compress the user-data especially because it could be limited to [~16384 bytes](http://cloudinit.readthedocs.io/en/latest/topics/format.html#gzip-compressed-content) with some providers.
In all these cases RancherOS won't be able to load and apply the cloud-config, because it makes no attempt to uncompress gzipped streams.

**Comment by niusmallnan** [comment]
@stffabi can you make a PR？

**Comment by greut** [comment]
things are there to do so... where does it infer the encoding?
https://github.com/rancher/os/blob/master/config/cloudinit/config/decode.go

**Comment by stffabi** [comment]
@greut that code is currently only used for the `write_files` section of a cloud-init file and for the vmware provider. The vmware provider is the only one which currently supports different encodings by providing the [cloud-init.data.encoding](https://rancher.com/docs/os/v1.2/en/running-rancheros/cloud/vmware-esxi/) guest-info property.
I think in this case we should detect if the content is gzipped by looking at the magic bytes header, because when downloading the content from the user-data server in most cases we won't get a correct mime-type or content-encoding.

@niusmallnan I can provide a PR to detect gzip content and decompress it.

**Comment by greut** [comment]
@stffabi good! I have to find my way around this code. Best of luck!

**Comment by stffabi** [comment]
@niusmallnan PR #2393  has been created.

**Comment by Jason-ZW** [comment]
@stffabi  PR #2393 LGTM.
I'm using the `aws cli` to verify the gzip `user-data`, steps as below:
1. create an yaml file:
```
#cloud-config
write_files:
 - path: /home/rancher/vpn_server.config
   permissions: "0644"
   owner: root
   content: test
```
2. using `gzip` compress the `boot.yml`:
`gzip boot.yml`
3. run `aws-cli` to create aws instance:
```
aws ec2 run-instances --image-id ami-a3383149 --region eu-west-1 --instance-type t2.small --security-group-ids sg-35b2634d --key-name xxxxx \
  --user-data fileb:///home/ubuntu/boot.yml.gz \
  --disable-api-termination \
  --instance-initiated-shutdown-behavior stop
```

After that I can see the `vpn_server.config` file in the ros instance:
```
test
```

**Comment by rootwuj** [comment]
This issue has been fixed in rancheros 1.4.0-rc2.

**Comment by niusmallnan** [comment]
Backported to v1.4.1.
